File: backend/app/core/context_manager.py
import json
import shutil
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional
from app.core.config import settings


class ContextManager:
    def __init__(self, project_id: str):
        self.project_id = project_id
        self.context_dir = Path(settings.storage.context_dir)
        self.context_file = self.context_dir / f"{project_id}.json"
        self.history_dir = self.context_dir / "history" / project_id
        
        self.context_dir.mkdir(parents=True, exist_ok=True)
        self.history_dir.mkdir(parents=True, exist_ok=True)
    
    def initialize(self, novel_text: str) -> Dict[str, Any]:
        context = {
            "meta": {
                "project_id": self.project_id,
                "version": 1,
                "created_at": datetime.utcnow().isoformat(),
                "novel_text": novel_text
            },
            "status": "parsing",
            "style": settings.project.default_style,
            "characters": {},
            "scenes": []
        }
        
        self._save_context(context)
        return context
    
    def load(self) -> Dict[str, Any]:
        if not self.context_file.exists():
            raise FileNotFoundError(f"Context file not found for project: {self.project_id}")
        
        with open(self.context_file, "r", encoding="utf-8") as f:
            return json.load(f)
    
    def update(self, updates: Dict[str, Any], create_snapshot: bool = True) -> Dict[str, Any]:
        if create_snapshot:
            self._create_snapshot()
        
        context = self.load()
        context["meta"]["version"] += 1
        context["meta"]["updated_at"] = datetime.utcnow().isoformat()
        
        for key, value in updates.items():
            if key in context:
                if isinstance(context[key], dict) and isinstance(value, dict):
                    context[key].update(value)
                else:
                    context[key] = value
            else:
                context[key] = value
        
        self._save_context(context)
        return context
    
    def update_character(self, character_name: str, field: str, value: Any) -> Dict[str, Any]:
        context = self.load()
        
        if character_name not in context["characters"]:
            context["characters"][character_name] = {}
        
        context["characters"][character_name][field] = value
        context["meta"]["version"] += 1
        context["meta"]["updated_at"] = datetime.utcnow().isoformat()
        
        self._save_context(context)
        return context
    
    def update_scene(self, scene_id: str, field: str, value: Any) -> Dict[str, Any]:
        context = self.load()
        
        for scene in context["scenes"]:
            if scene["id"] == scene_id:
                if field == "status" or field == "error":
                    scene[field] = value
                elif field in scene.get("assets", {}):
                    if "assets" not in scene:
                        scene["assets"] = {}
                    scene["assets"][field] = value
                else:
                    scene[field] = value
                break
        
        context["meta"]["version"] += 1
        context["meta"]["updated_at"] = datetime.utcnow().isoformat()
        
        self._save_context(context)
        return context
    
    def get_character(self, character_name: str) -> Optional[Dict[str, Any]]:
        context = self.load()
        return context["characters"].get(character_name)
    
    def get_scene(self, scene_id: str) -> Optional[Dict[str, Any]]:
        context = self.load()
        for scene in context["scenes"]:
            if scene["id"] == scene_id:
                return scene
        return None
    
    def rollback(self, version: int) -> Dict[str, Any]:
        snapshot_file = self.history_dir / f"v{version}.json"
        if not snapshot_file.exists():
            raise FileNotFoundError(f"Snapshot version {version} not found")
        
        shutil.copy(snapshot_file, self.context_file)
        return self.load()
    
    def _save_context(self, context: Dict[str, Any]):
        with open(self.context_file, "w", encoding="utf-8") as f:
            json.dump(context, f, indent=2, ensure_ascii=False)
    
    def _create_snapshot(self):
        if not self.context_file.exists():
            return
        
        context = self.load()
        version = context["meta"]["version"]
        snapshot_file = self.history_dir / f"v{version}.json"
        
        shutil.copy(self.context_file, snapshot_file)

File: frontend/lib/utils/context-manager.ts
import fs from 'fs/promises';
import path from 'path';
import type { ProjectContext, Character, Scene } from '../types';

export class ContextManager {
  private projectId: string;
  private contextDir: string;
  private contextPath: string;
  private historyDir: string;

  constructor(projectId: string) {
    this.projectId = projectId;
    this.contextDir = path.join(process.cwd(), 'context', projectId);
    this.contextPath = path.join(this.contextDir, 'context.json');
    this.historyDir = path.join(this.contextDir, 'history');
  }

  async ensureDirectories(): Promise<void> {
    await fs.mkdir(this.contextDir, { recursive: true });
    await fs.mkdir(this.historyDir, { recursive: true });
  }

  async initialize(characters: Record<string, Character>, scenes: Scene[]): Promise<ProjectContext> {
    await this.ensureDirectories();

    const context: ProjectContext = {
      meta: {
        version: 1,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      },
      status: 'text_parsed',
      characters,
      scenes,
    };

    await this.save(context);
    return context;
  }

  async load(): Promise<ProjectContext> {
    try {
      const data = await fs.readFile(this.contextPath, 'utf-8');
      return JSON.parse(data);
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        throw new Error('Project not found');
      }
      throw error;
    }
  }

  async save(context: ProjectContext): Promise<void> {
    await this.ensureDirectories();

    const existingContext = await this.load().catch(() => null);
    if (existingContext) {
      const historyFile = path.join(
        this.historyDir,
        `v${existingContext.meta.version}.json`
      );
      await fs.writeFile(historyFile, JSON.stringify(existingContext, null, 2));
    }

    context.meta.version = existingContext ? existingContext.meta.version + 1 : 1;
    context.meta.updated_at = new Date().toISOString();

    await fs.writeFile(this.contextPath, JSON.stringify(context, null, 2));
  }

  async update(updates: Partial<ProjectContext>): Promise<ProjectContext> {
    const context = await this.load();
    const updatedContext = { ...context, ...updates };
    await this.save(updatedContext);
    return updatedContext;
  }

  async updateStatus(status: string): Promise<void> {
    const context = await this.load();
    context.status = status;
    await this.save(context);
  }

  async updateCharacter(
    characterName: string,
    field: keyof Character,
    value: any
  ): Promise<void> {
    const context = await this.load();
    if (!context.characters[characterName]) {
      throw new Error(`Character ${characterName} not found`);
    }
    (context.characters[characterName] as any)[field] = value;
    await this.save(context);
  }

  async updateScene(sceneId: string, field: keyof Scene, value: any): Promise<void> {
    const context = await this.load();
    const scene = context.scenes.find((s) => s.id === sceneId);
    if (!scene) {
      throw new Error(`Scene ${sceneId} not found`);
    }
    (scene as any)[field] = value;
    await this.save(context);
  }

  async rollback(version: number): Promise<ProjectContext> {
    const historyFile = path.join(this.historyDir, `v${version}.json`);
    try {
      const data = await fs.readFile(historyFile, 'utf-8');
      const historicalContext = JSON.parse(data);
      await this.save(historicalContext);
      return historicalContext;
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        throw new Error(`Version ${version} not found`);
      }
      throw error;
    }
  }

  async listHistoryVersions(): Promise<number[]> {
    try {
      const files = await fs.readdir(this.historyDir);
      return files
        .filter((f) => f.startsWith('v') && f.endsWith('.json'))
        .map((f) => parseInt(f.slice(1, -5)))
        .sort((a, b) => b - a);
    } catch (error) {
      return [];
    }
  }
}
